vms/evm/predicate/results_test.go

```diff
@@ -4,227 +4,413 @@
 				},
 			},
 		},
 		{
-			name: "multiple txs mixed results",
-			results: map[common.Hash]PrecompileResults{
-				{1}: {
-					{2}: set.NewBits(1, 2, 3),
-				},
-				{2}: {
-					{2}: set.NewBits(1, 2, 3),
-					{3}: set.NewBits(3, 2, 1),
+			name: "delete_tx",
+			start: BlockResults{
+				txHash: PrecompileResults{
+					{3}: set.NewBits(1, 2, 3),
 				},
-				{3}: {},
 			},
+			toSet: PrecompileResults{},
+			want:  BlockResults{},
 		},
 	}
-
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			req := require.New(t)
-			// Build bytes for the encoded representation directly
-			enc := encodedBlockResults{TxResults: make(map[common.Hash]map[common.Address][]byte, len(tt.results))}
-			for tx, addrToBits := range tt.results {
-				addrToBytes := make(map[common.Address][]byte, len(addrToBits))
-				for addr, bits := range addrToBits {
-					addrToBytes[addr] = bits.Bytes()
-				}
-				enc.TxResults[tx] = addrToBytes
-			}
-			bytes, err := resultsCodec.Marshal(version, &enc)
-			req.NoError(err)
-
-			got, err := ParseBlockResults(bytes)
-			req.NoError(err)
-			req.Equal(BlockResults{TxResults: tt.results}, got)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			test.start.Set(txHash, test.toSet)
+			require.Equal(t, test.want, test.start)
 		})
 	}
 }
 
-func TestBlockResultsGetSet(t *testing.T) {
-	type setArgs struct {
-		tx      common.Hash
-		results PrecompileResults
-	}
-	type getArgs struct {
-		tx       common.Hash
-		addr     common.Address
-		wantBits set.Bits
-		wantOK   bool
-	}
-	type vecCase struct {
-		name    string
-		sets    []setArgs
-		gets    []getArgs
-		wantLen int
-	}
-
-	tests := []vecCase{
+func TestBlockResultsBytes(t *testing.T) {
+	tests := []struct {
+		name  string
+		input BlockResults
+		want  []byte
+	}{
 		{
-			name:    "zero value no sets",
-			sets:    nil,
-			gets:    []getArgs{{tx: common.Hash{1}, addr: common.Address{2}, wantBits: set.Bits{}, wantOK: false}},
-			wantLen: 0,
+			name:  "nil",
+			input: nil,
+			want: []byte{
+				// codecID
+				0x00, 0x00,
+				// results length
+				0x00, 0x00, 0x00, 0x00,
+			},
 		},
 		{
-			name: "single set and get",
-			sets: []setArgs{{
-				tx: common.Hash{1},
-				results: PrecompileResults{
-					common.Address{2}: set.NewBits(1, 2, 3),
+			name:  "empty",
+			input: make(BlockResults),
+			want: []byte{
+				// codecID
+				0x00, 0x00,
+				// results length
+				0x00, 0x00, 0x00, 0x00,
+			},
+		},
+		{
+			name: "single_tx_single_result",
+			input: BlockResults{
+				{1}: {
+					{2}: set.NewBits(1, 2, 3),
 				},
-			}},
-			gets: []getArgs{{
-				tx:       common.Hash{1},
-				addr:     common.Address{2},
-				wantBits: set.NewBits(1, 2, 3),
-				wantOK:   true,
-			}},
-			wantLen: 1,
-		},
-		{
-			name: "multiple txs and gets",
-			sets: []setArgs{
-				{
-					tx: common.Hash{1},
-					results: PrecompileResults{
-						common.Address{2}: set.NewBits(1, 2, 3),
-					},
+			},
+			want: []byte{
```

Comment:
Explicitly locking in the serialization format makes me a lot more comfortable to make changes here.